sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp

```diff
@@ -125,6 +125,11 @@ namespace Azure { namespace Core { namespace Cryptography {
   private:
     void* m_md5Context;
 
+    // Delete the copy and move constructors, along with the assignment operator.
+    Md5Hash(Md5Hash const&) = delete;
+    Md5Hash(Md5Hash const&&) = delete;
+    void operator=(Md5Hash const&) = delete;
```

Comment:
@CaseyCarter I would like to understand better the implications of explicitly deleting such ctors/operators, and when I need to use `const&&` vs just `&&`.

Also, are these "deletions" typically private?

Reply:
IIRC https://www.youtube.com/watch?v=vLinb2fgkHk&feature=emb_logo&ab_channel=InsideBloomberg is a good overview of move semantics and how explicit deletions suppress various of the special member functions (SMFs).

Deleted SMFs are typically not private, but with deletions it makes no difference. ("Code outside this class may not call this function which is uncallable in any case.")

`Md5Hash const&&` is a slightly more permissive parameter type than `Md5Hash &&` since the former can bind to `const Md5Hash` rvalues which the latter cannot. That fact also makes no appreciable difference in this case: if the move constructor parameter were plain `Md5Hash&&`, calls with such arguments invoke the deleted copy constructor instead of the deleted move constructor.

It's peculiar that this class explicitly deletes its move constructor but allows the move assignment operator to be deleted implicitly (since there's a user-declared copy assignment operator, the move assignment operator is implicitly deleted.) I'd expect to see either the maximally explicit "delete all four move/copy constructors/assignment operators" or the more terse "delete the copy operations and let the moves be implicitly deleted."

Reply:
>  I'd expect to see either the maximally explicit "delete all four move/copy constructors/assignment operators" or the more terse "delete the copy operations and let the moves be implicitly deleted."

So, if I am understanding this correctly, are you suggesting to apply one of these two options:
1) Either add `void operator=(Md5Hash const&&) = delete;` OR
2) Remove `Md5Hash(Md5Hash const&&) = delete;`

Also, since this class has a destructor, I can't follow the rule of 0, but I don't know how to follow the rule of 3 here:
> A copy constructor, a copy assignment operator, no move constructor, no move assignment operator, and a destructor.

Should I be adding a copy ctor/assignment instead of deleting it to adhere to the guideline?

Reply:
> > I'd expect to see either the maximally explicit "delete all four move/copy constructors/assignment operators" or the more terse "delete the copy operations and let the moves be implicitly deleted."
> 
> So, if I am understanding this correctly, are you suggesting to apply one of these two options:
> 
> 1. Either add `void operator=(Md5Hash const&&) = delete;` OR
> 2. Remove `Md5Hash(Md5Hash const&&) = delete;`

Yes, that's the general idea. I'd prefer the removal, since `void operator=` (vs. `Md5Hash& operator=`) and `(meow const&&)` (vs. `meow&&`) are both weird enough to make someone scrutinize the code to figure out why they're written that way.

> Also, since this class has a destructor, I can't follow the rule of 0, but I don't know how to follow the rule of 3 here:
> 
> > A copy constructor, a copy assignment operator, no move constructor, no move assignment operator, and a destructor.
> 
> Should I be adding a copy ctor/assignment instead of deleting it to adhere to the guideline?

I think the guideline's "should implement" is intended to admit `= default` and `= delete` as "implementation". In this case, we'd follow the rule of 3 by defining a destructor, and defining deleted copy constructor and copy assignment operations. Actually, having read through most of the file now I think it would make more sense to delete `Hash`'s copies so all classes that derive from `Hash` are non-copyable by default.

Reply:
> I think the guideline's "should implement" is intended to admit `= default` and `= delete` as "implementation". In this case, we'd follow the rule of 3 by defining a destructor, and defining deleted copy constructor and copy assignment operations. 

That is good to know. @RickWinter we should add this as a clarification to the guidelines, and include an example so folks reading it in the future can benefit from it as well.

> Actually, having read through most of the file now I think it would make more sense to delete `Hash`'s copies so all classes that derive from `Hash` are non-copyable by default.

👍

Reply:
why not just put the `std::string` and handle in primate members?

Of course we still need these overload/definition of constructor and assignment operator.

Reply:
I don't follow what you mean @Jinming-Hu .

Reply:
instead of 

```
private:
    void* m_md5Context;
```

```
private:
    std::string buffer;
    BCRYPT_HASH_HANDLE hashHandle;
```

Reply:
Because the member types would be different based on the OS (for example, it's `MD5_CTX` on non-Windows), and I didn't really want to expose bcrypt.h/openssl in the public header.

Reply:
> Actually, having read through most of the file now I think it would make more sense to delete `Hash`'s copies so all classes that derive from `Hash` are non-copyable by default.

@CaseyCarter, would I need to define a default `Hash()` ctor too? If I just have those methods deleted, I get an error on the ctor of the derived type, i.e. `Md5Hash`:
![image](https://user-images.githubusercontent.com/6527137/107466789-84b8f380-6b19-11eb-89b1-05d97f7dd32e.png)

Reply:
Yes; if we declare any kind of constructor the default constructor is not implicitly declared. So once we declare the deleted copy constructor in `Hash`, we'll need to define the default constructor explicitly (`Hash() = default;`).